stdlib/public/core/StringStorage.swift

```diff
@@ -16,59 +16,103 @@ import SwiftShims
+  str._copyUTF16CodeUnits(
+    into: UnsafeMutableBufferPointer(start: buffer, count: range.count),
+    range: range)
+}
 
-    let range = Range(
-      uncheckedBounds: (aRange.location, aRange.location+aRange.length))
-    let str = asString
-    str._copyUTF16CodeUnits(
-      into: UnsafeMutableBufferPointer(start: buffer, count: range.count),
-      range: range)
+@inline(never) //hide the shim call so we can use @_effects
+@_effects(readonly)
+private func _isNSString(_ str:AnyObject) -> UInt8 {
+  return _swift_stdlib_isNSString(str)
+}
+
+@inline(__always)
+@_effects(readonly)
+private func _isEqual<T:_AbstractStringStorage>(_ this:T, _ other:AnyObject?)
+  -> Int8 {
+  guard let other = other else {
+    return 0
   }
 
-  @objc(_fastCharacterContents)
-  final internal func _fastCharacterContents() -> UnsafePointer<UInt16>? {
-    return nil
+  if this === other {
+    return 1
   }
+ 
+  let ourGuts = this.asGuts
+  defer { _fixLifetime(ourGuts) }
 
-  @objc(_fastCStringContents)
-  final internal func _fastCStringContents() -> UnsafePointer<CChar>? {
-    if let native = self as? _StringStorage, native.isASCII {
-      return native.start._asCChar
+  //Handle the case where both strings were bridged from Swift.
+  //We can't use String.== because it doesn't match NSString semantics.
+  if let otherGuts = (other as? _AbstractStringStorage)?.asGuts {
+    if otherGuts.count != ourGuts.count {
+      return 0
     }
+    return ourGuts.withFastUTF8 { ourBytes in
+      defer {
+        _fixLifetime(otherGuts)
+      }
+      return otherGuts.withFastUTF8 { otherBytes in
+        return (ourBytes.baseAddress == otherBytes.baseAddress ||
+          (memcmp(ourBytes.baseAddress!, otherBytes.baseAddress!, ourBytes.count) == 0)) ? 1 : 0
+      }
+    }
+  }
+  
+  //we're allowed to crash, but for compatibility reasons NSCFString allows non-strings here
+  if _isNSString(other) != 1 {
+    return 0
+  }
+  
+  //At this point we've proven that it is an NSString of some sort, but not one of ours
+  if this.length != _stdlib_binary_CFStringGetLength(other) {
+    return 0
+  }
+  
+  defer {
+    _fixLifetime(other)
+  }
 
-    // TODO(String performance): Check for nul-terminated shared strings, which
-    // could be from bridged literals two/from ObjC (alternatively: reconsider
-    // our bridging model for literals).
-
-    return nil
+  //CFString will only give us ASCII bytes here, but that's fine
+  //We already handled non-ASCII UTF8 strings earlier since they're Swift
+  if let otherBytes = _cocoaUTF8Pointer(other) {
+    return ourGuts.withFastUTF8 { ourBytes in
+      return (ourBytes.baseAddress == otherBytes ||
+        (memcmp(ourBytes.baseAddress!, otherBytes, ourBytes.count) == 0)) ? 1 : 0
+    }
   }
+  
+  return _cocoaStringCompare(this, other) == 0 ? 1 : 0
+}
+
+extension _AbstractStringStorage {
+  //in UTF16 code units
+  @objc(length) internal var length: Int { @_effects(readonly) get { Builtin.unreachable() } }
```

Comment:
I should make sure this isn't generating an objc category. I put it here because it's in the #if _runtime(_ObjC), but if we're paying anything for it I should just add a new conditional block above instead.

Reply:
It probably is generating a category. We ought to collect all the declarations ahead of time, or at least the ones in the same file, but I don't think we do.